### PR TITLE
Gh 1289/migrate push notification registration to client gateway

### PR DESCRIPTION
**Pull request by MouazAlzahabi**

### CODE REVIEW

File: Multisig/App/RemoteNotificationHandler.swift [hunk 1]
```diff
@@ -232,24 +197,44 @@ class RemoteNotificationHandler {
...
-                try SafeTransactionService.unregister(deviceID: self.storedDeviceID!,
-                                                      address: address,
-                                                      networkId: networkId)
-            } catch {
-                logError("Failed to unregister device", error)
-            }
-        }
+        App.shared.clientGatewayService.unregister(deviceID: self.storedDeviceID!,
+                                              address: address,
+                                              networkId: networkId)
```
Comment (DmitryBespalov) [inline Multisig/App/RemoteNotificationHandler.swift]: indentation seems off

File: Multisig/App/RemoteNotificationHandler.swift [hunk 2]
```diff
@@ -232,24 +197,44 @@ class RemoteNotificationHandler {
...
+        }
+
+        do {
+            let appConfig = App.configuration.app
+            let timestamp = String(format: "%.0f", Date().timeIntervalSince1970)
+
+            var safeRegistration: [SafeRegistration] = []
+            for networkSafes in Network.networkSafes() {
+                let safes = networkSafes.safes.compactMap { $0.address }.compactMap { Address($0) }.map { $0.checksummed }.sorted()
+                let signResult = try Self.sign(safes: safes, deviceID: deviceID, token: token, timestamp: timestamp)
```
Comment (DmitryBespalov) [inline Multisig/App/RemoteNotificationHandler.swift]: do we need to check for empty `safes` and `continue` in that case?

File: Multisig/Data/Services/Safe Client Gateway Service/RegisterNotificationTokenRequest.swift [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+//
+//  Created by Moaaz on 8/5/20.
+//  Copyright © 2020 Gnosis Ltd. All rights reserved.
+//
+
+import Foundation
+import Web3
+
+struct RegisterNotificationTokenRequest: JSONRequest {
+    //var deviceData: DeviceData
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Safe Client Gateway Service/RegisterNotificationTokenRequest.swift]: let's remove if not needed

File: Multisig/Data/Services/Safe Client Gateway Service/UnregisterNotificationTokenRequest.swift [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import Foundation
+
+struct UnregisterNotificationTokenRequest: JSONRequest {
+    let deviceID: String
+    let address: String
+    let networkId: String
+    var httpMethod: String { return "DELETE" }
+    var urlPath: String { return "/v1/chains/\(networkId)/notifications/devices/\(deviceID)/safes/\(address)/" }
+    typealias ResponseType = EmptyResponse
+    init(deviceID: String, address: Address, networkId: String) {
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Safe Client Gateway Service/UnregisterNotificationTokenRequest.swift:21]: some blank lines would help

**Review by DmitryBespalov — APPROVED** [review]
small comments, otherwise ok

File: Multisig/App/RemoteNotificationHandler.swift (hunk 2 above)
Comment (MouazAlzahabi) [reply Multisig/App/RemoteNotificationHandler.swift]: I have made same check in different way, please check

**Review by DmitryBespalov — APPROVED** [review]

File: Multisig/App/RemoteNotificationHandler.swift [hunk 5]
```diff
@@ -206,18 +173,17 @@ class RemoteNotificationHandler {
...
-    static func sign(safes: [String], deviceID: String, token: String, timestamp: String? = nil) throws -> (preimage: String, hash: String, timestamp: String, signatures: [String])? {
+    static func sign(safes: [String], deviceID: String, token: String, timestamp: String) throws -> (preimage: String, hash: String, signatures: [String])? {
         // sign the registration data by each private key.
 
         let privateKeys = try KeyInfo.keys(types: [.deviceImported, .deviceGenerated])
             .compactMap { try $0.privateKey() }
 
-        guard !privateKeys.isEmpty else {
+        guard !privateKeys.isEmpty,
+              !safes.isEmpty else {
```
Comment (sche) [inline Multisig/App/RemoteNotificationHandler.swift:183]: why do we need this check?

**Review by sche — APPROVED** [review]

File: Multisig/App/RemoteNotificationHandler.swift (hunk 5 above)
Comment (MouazAlzahabi) [reply Multisig/App/RemoteNotificationHandler.swift:183]: because this code is being executed when we import safe without having owner imported

**Pull request merged**